src/metadatad.cc

```diff
@@ -14,28 +14,69 @@
  * limitations under the License.
  **/
 
+#include <csignal>
+#include <cstdlib>
+#include <initializer_list>
+#include <iostream>
+
 #include "agent.h"
 #include "configuration.h"
 #include "docker.h"
 #include "instance.h"
 #include "kubernetes.h"
 
+namespace google {
+namespace {
+
+struct CleanupState {
+  CleanupState(
+      std::initializer_list<MetadataUpdater*> updaters_, MetadataAgent* server_)
+      : updaters(updaters_), server(server_) {}
+  std::vector<MetadataUpdater*> updaters;
+  MetadataAgent* server;
+};
+const CleanupState* cleanup_state;
+
+}  // namespace
+}  // google
+
+extern "C" [[noreturn]] void handle_sigterm(int signum) {
+  std::cerr << "Caught SIGTERM; shutting down" << std::endl;
+  std::cerr << "Stopping server" << std::endl;
+  google::cleanup_state->server->stop();
+  std::cerr << "Stopping updaters" << std::endl;
+  for (google::MetadataUpdater* updater : google::cleanup_state->updaters) {
+    updater->stop();
+  }
+  std::cerr << "Exiting" << std::endl;
+  std::exit(128 + signum);
+}
+
 int main(int ac, char** av) {
   google::Configuration config;
   int parse_result = config.ParseArguments(ac, av);
   if (parse_result) {
     return parse_result < 0 ? 0 : parse_result;
   }
 
+  std::mutex server_wait_mutex;
+  server_wait_mutex.lock();
```

Comment:
Is this line required given the lock guard in L84?
Alternatively, should this lock be released after server.start()?

Reply:
The lock guard was supposed to deadlock this thread to prevent it from going into the destructor. However, with the suggestion from @bmoyles0117, this is no longer necessary.